hypothesis/lms
# Fix `static_url()` in Celery tasks

When trying to move the email digests template into Python my new Jinja2 template for the email body has to call `static_url` to generate the URL for an image file that's used in the email (this image file will now be hosted as a static asset in the LMS app). I found that `static_url` returned URLs like `http://localhost/static/foo.png` without the port: Pyramid methods like `static_url()`, `route_url()` etc don't work when called by Celery tasks because they don't have the HTTP host information (`localhost:8001` in dev) that would normally come from the HTTP request's `Host` header.

This PR fixes this by injecting the correct `HTTP_HOST` variable into the WSGI environment in Celery processes. This is the same WSGI environment variable that would normally be set based on the `Host` header of the HTTP request when responding to an HTTP request rather than running a Celery task.

I'm not sure if this is the best/right way to do this with Pyramid but it works.

This requires each deployment of the app (dev, QA and prod) to have an `HTTP_HOST` envvar set correctly.

This replaces a previous solution where we had a `WEB_APP_URL`/`web_app_url` envvar/setting and we passed a wrapped `route_url()` method to `EmailUnsubscribeService`:

```python
route_url=partial(
    request.route_url, _app_url=request.registry.settings["web_app_url"]
)
```

This only worked for `route_url` and only for `EmailUnsubscribeService`. The solution in this commit should fix both `route_url()` and `static_url()` and should fix them for all code called by Celery tasks.

The previous `WEB_APP_URL` envvar (e.g. `https://qa-lms.hypothies.is`) will have to be replaced by the new `HTTP_HOST` envvar (just `qa-lms.hypothes.is`) in all environments.

### Testing

1. Log in to https://hypothesis.instructure.com/ as `eng+canvasteacher@hypothes.is` and launch [localhost (make devdata) HTML Assignment](https://hypothesis.instructure.com/courses/125/assignments/873). This is to make sure that you have a record of the instructor in your DB.

2. Log out and log back in as `eng+canvasstudent@hypothes.is`. Launch the same assignment again and create some annotations.

3. Go to http://localhost:8001/admin/email, enter your email address, change the since and until dates so that they cover the time when you created the annotations, and click send.

4. Look for the `unsubscribe_url` that was printed in the terminal, it should be the correct URL including the port:

   ```
   {'name': 'unsubscribe_url', 'content': 'http://localhost:8001/email/unsubscribe?token=***'}
   ```
